api/v1alpha1/contour_types.go

```diff
@@ -91,11 +102,160 @@ type NamespaceSpec struct {
+	//
+	// * LoadBalancerService
+	//
+	// In this configuration, network endpoints for Envoy use container networking.
+	// A Kubernetes LoadBalancer Service is created to publish Envoy network
+	// endpoints. The Service uses port 80 to publish Envoy's HTTP network endpoint
+	// and port 443 to publish Envoy's HTTPS network endpoint.
+	//
+	// See: https://kubernetes.io/docs/concepts/services-networking/service/#loadbalancer
+	//
+	// * NodePortService
+	//
+	// Publishes Envoy network endpoints using a Kubernetes NodePort Service.
+	//
+	// In this configuration, Envoy network endpoints use container networking. A Kubernetes
+	// NodePort Service is created to publish the network endpoints.
+	//
+	// See: https://kubernetes.io/docs/concepts/services-networking/service/#nodeport
+	//
+	// +unionDiscriminator
+	// +kubebuilder:default=LoadBalancerService
+	Type NetworkPublishingType `json:"type,omitempty"`
+
+	// loadBalancer holds parameters for the load balancer. Present only if type is
+	// LoadBalancerService.
+	//
+	// If unspecified, defaults to an external Classic AWS ELB.
+	//
+	LoadBalancer LoadBalancerStrategy `json:"loadBalancer,omitempty"`
+
+	// ContainerPorts is a list of container ports to expose from the Envoy container(s).
+	// Exposing a port here gives the system additional information about the network
+	// connections the Envoy container uses, but is primarily informational. Not specifying
+	// a port here DOES NOT prevent that port from being exposed by Envoy. Any port which is
+	// listening on the default "0.0.0.0" address inside the Envoy container will be accessible
+	// from the network. Names and port numbers must be unique in the list container ports. A
+	// minimum of two ports must be specified, one named "http" for Envoy's insecure service
+	// and one named "https" for Envoy's secure service.
+	//
+	// +kubebuilder:validation:MinItems=2
+	// +kubebuilder:validation:MaxItems=6
+	// +kubebuilder:default={{name: http, portNumber: 8080}, {name: https, portNumber: 8443}}
+	ContainerPorts []ContainerPort `json:"containerPorts,omitempty"`
+}
+
+// EndpointPublishingType is a way to publish network endpoints.
+// +kubebuilder:validation:Enum=LoadBalancerService;NodePortService
+type NetworkPublishingType string
+
+const (
+	// LoadBalancerService publishes a network endpoint using a Kubernetes LoadBalancer
+	// Service.
+	LoadBalancerServicePublishingType NetworkPublishingType = "LoadBalancerService"
+
+	// NodePortService publishes a network endpoint using a Kubernetes NodePort Service.
+	NodePortServicePublishingType NetworkPublishingType = "NodePortService"
+)
+
+// LoadBalancerStrategy holds parameters for a load balancer.
+type LoadBalancerStrategy struct {
+	// Scope indicates the scope at which the load balancer is exposed.
+	// Possible values are "External" and "Internal".
+	//
+	// +kubebuilder:default=External
+	Scope LoadBalancerScope `json:"scope,omitempty"`
```

Comment:
Is the operator going to take on specific implementations (e.g. https://kubernetes.io/docs/concepts/services-networking/service/#internal-load-balancer). 

If the goal isn't to take those on then possibly a new `ProviderSetting` (or something) are a better spot to allow users to define those specific exceptions. We might need both to cover anyone who might need specific annotations but aren't yet supported in the operator.

Reply:
@stevesloka yes, the operator will support provider-specific LB settings through `ProviderParameters`, i.e. LB connection tuning. All the major cloud providers support internal/external LB scoping, so I think it's appropriate to define scope outside `ProviderParameters`.